File: backend/domains/user/handlers/get_user_profile.handler.go
package handlers

import (
	"context"

	"github.com/cloudwego/hertz/pkg/app"
	"github.com/cloudwego/hertz/pkg/common/utils"
	"github.com/erweixin/go-genai-stack/backend/infrastructure/middleware"
)

// GetUserProfileHandler 获取用户资料 Handler
//
// 对应用例：GetUserProfile
// HTTP Method: GET
// HTTP Path: /api/users/me
// 认证：必需
//
// Handler 职责（瘦层）：
//   - 从 JWT Token 中获取用户 ID
//   - 转换 DTO（使用转换层）
//   - 调用 Domain Service
//   - 转换响应（使用转换层）
func (deps *HandlerDependencies) GetUserProfileHandler(ctx context.Context, c *app.RequestContext) {
	// 1. 从认证中间件获取用户 ID
	userID, err := middleware.RequireUserID(c)
	if err != nil {
		c.JSON(401, utils.H{
			"error": "UNAUTHORIZED",
			"message": "未授权访问",
		})
		return
	}

	// 2. 转换为 Domain Input（使用转换层）
	input := toGetUserProfileInput(userID)

	// 3. 调用 Domain Service
	output, err := deps.userService.GetUserProfile(ctx, input)
	if err != nil {
		handleDomainError(c, err)
		return
	}

	// 4. 转换为 HTTP 响应（使用转换层）
	c.JSON(200, toGetUserProfileResponse(output))
}

File: backend-nodejs/domains/user/handlers/get_user_profile.handler.ts
/**
 * GetUserProfile Handler
 * HTTP 适配层：处理获取用户资料的 HTTP 请求
 */

import type { FastifyRequest, FastifyReply } from 'fastify';
import type { HandlerDependencies } from './dependencies.js';
import {
  toGetUserProfileInput,
  toGetUserProfileResponse,
} from './converters.js';
import { requireUserId } from '../../../infrastructure/middleware/auth.js';
import { createContextFromRequest } from '../../../shared/types/context.js';

/**
 * GetUserProfile Handler
 * HTTP 适配层：处理获取用户资料的 HTTP 请求
 * 错误由全局错误处理中间件统一处理
 */
export async function getUserProfileHandler(
  deps: HandlerDependencies,
  req: FastifyRequest,
  reply: FastifyReply
): Promise<void> {
  // 1. 获取用户 ID（从 JWT Token 中提取）
  const userId = requireUserId(req);

  // 2. 转换为 Domain Input
  const input = toGetUserProfileInput(userId);

  // 3. 创建请求上下文
  const ctx = createContextFromRequest({ userId, ...req });

  // 4. 调用 Domain Service（错误会自动向上抛出，由全局错误处理中间件处理）
  const output = await deps.userService.getUserProfile(ctx, input);

  // 4. 转换为 HTTP 响应
  reply.code(200).send(toGetUserProfileResponse(output.user));
}

